File: 17. Letter Combinations of a Phone Number.py
class Solution(object):
    def __init__(self):
        self.visit = []
        self.mp = ['', '', 'abc', 'def', 'ghi', 'jkl', 'mno', 'pqrs', 'tuv', 'wxyz']
        
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        res = []
        self.visit = [False for x in range(len(digits))]
        self.dfs(digits, '', 0, res)
        return res
        
    def dfs(self, digits, num, depth, res):
        if depth == len(digits) and num:
            res.append(num)
            return
        i = len(num)
        while i < len(digits):
            if not self.visit[i]:
                self.visit[i] = True
                for x in self.mp[int(digits[i])]:
                    self.dfs(digits, num+x, depth+1, res)
                self.visit[i] = False
            i += 1
                    
s = Solution()
print s.letterCombinations('23')        

File: go/src/leetcode/17. Letter Combinations of a Phone Number/17. Letter Combinations of a Phone Number.go
package main

import "fmt"

var mp = map[uint8]string {
	'2' : "abc",
	'3' : "def",
	'4' : "ghi",
	'5' : "jkl",
	'6' : "mno",
	'7' : "pqrs",
	'8' : "tuv",
	'9' : "wxyz",
}

func letterCombinations(digits string) []string {
	ans := []string{}
	dfs(digits, 0, &[]uint8{}, &ans)
	return ans
}

func dfs(digits string, idx int, arr *[]uint8, ans *[]string)  {
	if len(*arr) > len(digits) {
		return
	}

	if idx == len(digits) {
		if len(*arr) != 0 && len(*arr) == len(digits) {
			*ans = append(*ans, string(*arr))
		}
		return
	}

	for i := idx; i < len(digits); i++ {
		for _, c := range mp[digits[i]] {
			*arr = append(*arr, uint8(c))
			dfs(digits, i+1, arr, ans)
			*arr = (*arr)[:len(*arr)-1]
		}
	}
}

func main() {
	fmt.Println(letterCombinations("23"))
	fmt.Println(letterCombinations(""))
	fmt.Println(letterCombinations("2"))
}
